### PR TITLE
Initial implementation of a Spark connector

### PR DESCRIPTION
**Pull request by n-riesco**

### CODE REVIEW

File: test/backend/datastores.livy.spec.js [hunk 1]
```diff
@@ -76,7 +76,8 @@ df.registerTempTable("ALCOHOL_CONSUMPTION_BY_COUNTRY_2010")
             'ALCOHOL_CONSUMPTION_BY_COUNTRY_2010';
 
         return query(`SELECT * FROM ${tableName}\nLIMIT 5`, connection).then(results => {
-            assert.deepEqual(results.rows, [
+            const roundedToSinglePrecision = results.rows.map(r => [r[0], Number(r[1].toPrecision(6))]);
```
Comment (n-riesco) [inline test/backend/datastores.livy.spec.js]: It looks like when the table was created in Spark, the source CSV file was imported as single-precision numbers. But when we query Spark/Livy, the results are reported as double-precision numbers.

@chriddyp Have you come across this issue with any of the other connectors?

Comment (chriddyp) [reply test/backend/datastores.livy.spec.js]: Hm, I haven't come across this although I haven't really investigated too deeply either

Comment (n-riesco) [reply test/backend/datastores.livy.spec.js]: @tarzzz @chriddyp Here's an untested explanation: I've just noticed that [the script to create the sample Hive table declares the column as `FLOAT`](https://github.com/plotly/plotly-database-connector/blob/2eae53de7347051328ea3cd6bd428609d765a90d/test/docker/hadoop-spark/create_hive_tables.py#L14).

**Comment by tarzzz** [comment]
@n-riesco @chriddyp  .. 

My only concern is with the workaround for float-precision in tests, but we can revert it later once test database is fixed (i.e. using decimal instead of float in its column)..   Good for now.. 💃

**Comment by chriddyp** [comment]
It looks like the build failed:
```
-----> Build failed
 !     Outdated Yarn lockfile
       Your application contains a Yarn lockfile (yarn.lock) which does not
       match the dependencies in package.json. This can happen if you use npm
       to install or update a dependency instead of Yarn.
       Please run the following command in your application directory and check
       in the new yarn.lock file:
       $ yarn install
       $ git add yarn.lock
       $ git commit -m "Updated Yarn lockfile"
       $ git push heroku master
    
       https://kb.heroku.com/why-is-my-node-js-build-failing-because-of-an-outdated-yarn-lockfile
 !     Push rejected, failed to compile Node.js app.
 !     Push failed
```

File: backend/persistent/datastores/livy.js [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    return newSession(connection);
+}
+
+export function tables(connection) {
+    const code = (connection.database) ?
+        `val r = plotlyContext.sql("show tables in ${connection.database}").collect\n%json r` :
+        `val r = plotlyContext.sql("show tables").collect\n%json r`;
+    return sendRequest(connection, code)
+        .then(json => {
+            if (!Array.isArray(json) || json.length <= 0) {
```
Comment (chriddyp) [inline backend/persistent/datastores/livy.js]: why `<`? can this just be `=== 0`?

File: backend/persistent/datastores/livy.js [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+function getClient(connection, attempts) {
+    const timeout = connection.timeout || 180;
+
+    attempts = attempts || 1
+    if (attempts++ > connection.timeout) new Error("livy: connect: timeout");
+
+    return getState(connection).then(state => {
+        return (state === 'idle') ?
+            connection :
+            wait(1000).then(function() {
```
Comment (chriddyp) [inline backend/persistent/datastores/livy.js]: 😻  love this `wait`

File: sample-storage/connections.yaml [hunk 4]
```diff
@@ -78,3 +78,11 @@
...
     port: 50000
     host: 35.184.35.183
+
+-
+    dialect: apache spark
+    id: apache spark-7b7bfa1e-cdfa-4001-b326-e0e170667021
+    timeout: 180
+    database: plotly
+    port: 8998
+    host: 104.155.170.7
```
Comment (chriddyp) [inline sample-storage/connections.yaml]: should there be another one in here for the sample IBM DB? 

It's not really documented, but `sample-storage` is used for the test heroku instance (heroku doesn't have a persistent file system, so we need to commit these credentials as part of the repo and then point the settings folder to this `sample-storage` folder to preload the available connections).

**Comment by chriddyp** [comment]
Looks good! Just a couple of small comments + the build, then 💃

**Comment by tarzzz** [comment]
@chriddyp The build should succeed after merging #175 and rebase, since it is related to yarn lockfile..

**Comment by chriddyp** [comment]
OK, I'll create a review app there. It looks like the review apps aren't being created automatically since this PR is based off of a fork.
Here's where you can create the review apps in heroku:
image

**Comment by chriddyp** [comment]
Could we add Spark and DB2 to the end of the list but right before Apache Drill? Right now, they are the first two items but they are probably less common than MySQL
image
but more common than Apache Drill

**Comment by chriddyp** [comment]
The Spark Instance "Had trouble connecting" but I'm assuming that this is an unrelated issue?
image

File: sample-storage/connections.yaml (hunk 4 above)
Comment (chriddyp) [reply sample-storage/connections.yaml]: NVM, it's already there :woman_facepalming:

File: backend/persistent/datastores/livy.js (hunk 2 above)
Comment (n-riesco) [reply backend/persistent/datastores/livy.js]: It's just a style thing. I'll change it to `===`.

**Comment by n-riesco** [comment]
Hi @chriddyp I've addressed the issues you raised. I've also added a commit to improve error handling from Spark.

The PR is ready for review again.

**Comment by chriddyp** [comment]
Great, thanks @n-riesco ! 💃  from me. @tarzzz ?

**Comment by tarzzz** [comment]
Yep, LGTM.. 💃 🎉 🎉

**Pull request merged**